### DISCUSSION TITLE
Dashboard graphics and enclosures

### DISCUSSION THREAD

**Discussion by steveava** [theyosh/TerrariumPI#791] [General]
Hi,

I have added a second enclosure to TP and seem to have lost graphs on the Dashboard as well as details of the second enclosure.

![image](https://user-images.githubusercontent.com/39738639/229348157-637d4432-2558-4377-8875-2d996e28535b.png)

Dashboard mode seems to be set correctly.

![image](https://user-images.githubusercontent.com/39738639/229348197-39d3d01d-e403-453a-9264-8ef543a82c9f.png)

 When changing to "Only graphs", I get the usual "Up Time, Power Usa.................. etc. but no graphs. If I change to "Only enclosures " I get this.

![image](https://user-images.githubusercontent.com/39738639/229349121-ec90dc7a-9265-4423-ad04-470b720f4402.png)

If I change back to "Graphs and enclosures I lose the graphs as well as the second enclosure details(top image),

I have a second RPI/TP that also doesn't display the graphs on the Dashboard but there is only one enclosure on that setup. 

Cant seem to find any other settings that would change Dashboard behaviour so just wondered if I have missed a setting somewhere.

Any help would be appreciated.
Thanks.

**Comment by theyosh** [comment]
Hmm, could you set it to show both graphs and enclosures. Then reload the page. And then look at the browser console. I suspect a javascript error.

https://appuals.com/open-browser-console/

And look for the 'console' where javascript errors can appear. If there are, please make a screenshot or copy the contents and post it here.

**Comment by steveava** [comment]
Ok,

 

This is what I get.

 

main.js:290 Uncaught (in promise) TypeError: Reduce of empty array with no initial value

    at Array.reduce (<anonymous>)

    at h2 (main.js:290:15745)

    at main.js:310:32167

    at Array.map (<anonymous>)

    at main.js:310:31930

    at main.js:31:7318

    at Generator.next (<anonymous>)

    at r (main.js:31:6174)

h2 @ main.js:290

(anonymous) @ main.js:310

(anonymous) @ main.js:310

(anonymous) @ main.js:31

r @ main.js:31

Promise.then (async)

Mo @ main.js:1

(anonymous) @ main.js:1

Nt @ main.js:1

(anonymous) @ main.js:310

Em @ main.js:24

(anonymous) @ main.js:27

Mm @ main.js:27

Promise.then (async)

Im @ main.js:27

qE @ main.js:27

(anonymous) @ main.js:27

(anonymous) @ main.js:290

Hc @ main.js:1

Promise.then (async)

Mo @ main.js:1

Hc @ main.js:1

Promise.then (async)

Mo @ main.js:1

(anonymous) @ main.js:1

Nt @ main.js:1

(anonymous) @ main.js:290

o @ main.js:27

n2 @ main.js:290

It @ main.js:27

a2 @ main.js:290

n8 @ main.js:558

It @ main.js:27

r8 @ main.js:562

(anonymous) @ main.js:562

(anonymous) @ main.js:562

**Comment by theyosh** [comment]
Hmm, not sure what is causing the error. So I tried to see if I could find something. Do if you do a `git pull` and reload the browser holding the shift key. Make sure you reload a fresh copy.

If that does not solves it, I need to think a bit more.

**Comment by steveava** [comment]
Done a Git pull then shift reload. Seem to get the same error and this is on the RPI with just one enclosure. Also tried Firefox and had the similar issue.

Chrome:
![image](https://user-images.githubusercontent.com/39738639/230029155-81bdcb1d-6aeb-4009-9b07-218ce5bd8a9e.png)

Firefox:
![image](https://user-images.githubusercontent.com/39738639/230031296-e21d1990-2d09-4d28-9e3e-a3d1e9e90011.png)

Also had this in Chrome under Issue Tab. not sure if it's related.
![image](https://user-images.githubusercontent.com/39738639/230029958-11ccf8a7-1fd9-44a0-abd9-8cb29715989e.png)

**Comment by steveava** [comment]
Just to add information and again not sure if its related, I also tried to make use of the calendar for veg nutrient levels and dates. I made an entry for feeding start date and although it was acepted, it didn't show in the calendar.

![image](https://user-images.githubusercontent.com/39738639/230036707-2e81e690-dc13-407d-addf-a4d1fb19adf9.png)

**Comment by steveava** [comment]
Think I may have discovered a setting that stops the graphs from displaying in Dashboard. I have two Enclosures with separate nutrient tanks with their own pump, heater and temperature sensors. The Exclude Average on the temperature sensors is turned on as the two systems are separately controlled. Switching them off seem to make a difference to how things are displayed in dashboard and the console error disappears. However this doesn't make both enclosures display details, only one seems to show up. And the calendar displays the same error regardless.

**Comment by theyosh** [comment]
I found the bug, and I updated the code. So do a `git pull` and reload the browser. The dashboard should now be working again.

And you second enclosure is correct, try clicking on the title ;)

**Comment by steveava** [comment]
Ah yes, clicking the title does show Enclosure details. Not sure why I didn't try that before!!

Done a git pull on one pi and it updated. Refreshed browser but it still doesn't show the graphs for each enclosure unless I enable the sensor  averaging.

The second pi is a little odd. Had this when trying a git pull.

![image](https://user-images.githubusercontent.com/39738639/230634750-582bb09d-de35-4645-aadc-d3fc3cc1eeb4.png)

**Reply by theyosh** [reply]
The git message can be ignored, as you are not adding code. So that should not interfere.

But your dashboard will only show graphs of average sensors values that are not excluded. So if you have for example 1 temperature sensor, and you will exclude that from average calculation, there will be no temperature graph on the dashboard.

Graphs per enclosure is not supported. I have thought about that, but I do not know how to do that nicely.

So the dashboard will show the average per sensor type. Despite which enclosure used.

**Comment by steveava** [comment]
Ok, understand.

Any thoughts on the calendar not displaying entries?

![image](https://user-images.githubusercontent.com/39738639/230639920-4b71d23c-1931-48b8-b8a9-f5c32bca29d0.png)

**Comment by theyosh** [comment]
Hmm, easiest is to run in debug mode. So make sure TerrariumPI is stopped: https://theyosh.github.io/TerrariumPI/faq/how-to-start-stop-restart-and-disable-enable-terrariumpi/

And then run it in debug mode: https://theyosh.github.io/TerrariumPI/faq/how-to-debug-terrariumpi/

Then you can try to load the calendar. I expect some errors showing up in the console. If you can post that here, I can have a look what is happening.

**Comment by steveava** [comment]
This is what I get from debug.

Traceback (most recent call last):
  File "/home/pi/TerrariumPI/venv/lib/python3.7/site-packages/bottle.py", line 876, in _handle
    return route.call(**args)
  File "/home/pi/TerrariumPI/venv/lib/python3.7/site-packages/bottle.py", line 1759, in wrapper
    rv = callback(*a, **ka)
  File "/home/pi/TerrariumPI/terrariumWebserver.py", line 88, in wrapper
    return func(*a, **ka)
  File "/home/pi/TerrariumPI/terrariumAPI.py", line 531, in calendar_list
    start = datetime.fromisoformat(start)
ValueError: Invalid isoformat string: '2023-03-26T00:00:00Z'

**Comment by theyosh** [comment]
Ok, do a `git pull` and restart TP4. That should fix it. The `Z` in the date time string is not expected. So that is now taken care off.

**Comment by steveava** [comment]
Tried to make a an entry for yesterday and you get the dialog box to fill out and save as before but it still doesn't show in calendar. The little icon at the top of the page shows a "1" in a small blue box but nothing on the calendar.

Get this in browser console which is a different error.

![image](https://user-images.githubusercontent.com/39738639/231135444-b1642696-e3f5-4b6b-8bd0-c26db3a570b8.png)

**Reply by theyosh** [reply]
This is again a server error. This needs again running in debug mode to see what the error is.

So can you run TP4 in debug again and try to load the calendar. I have no clue what is causing this all...

**Comment by steveava** [comment]
Looks like we get the same error refreshing the calendar and I also tried to add an entry. When you click the save button you get the same error.


/home/pi/TerrariumPI/venv/lib/python3.7/site-packages/bottle.py:3383: DeprecationWarning: Absolute template path names are deprecated.
  fname = self.search(name, self.lookup)
Traceback (most recent call last):
  File "/home/pi/TerrariumPI/venv/lib/python3.7/site-packages/bottle.py", line 876, in _handle
    return route.call(**args)
  File "/home/pi/TerrariumPI/venv/lib/python3.7/site-packages/bottle.py", line 1759, in wrapper
    rv = callback(*a, **ka)
  File "/home/pi/TerrariumPI/terrariumWebserver.py", line 88, in wrapper
    return func(*a, **ka)
  File "/home/pi/TerrariumPI/terrariumAPI.py", line 531, in calendar_list
    start = datetime.fromisoformat(start)
ValueError: Invalid isoformat string: '2023-03-26T00:00:00Z'

**Reply by theyosh** [reply]
This looks old code. 
https://github.com/theyosh/TerrariumPI/blob/main/terrariumAPI.py#L531

So could you do a `git pull` and restart? That should fix it. Not sure why you do not have the latest code.

**Comment by steveava** [comment]
Tried that several times already and still get this, although it did update some files earlier.

![image](https://user-images.githubusercontent.com/39738639/232587896-b9ebe73e-ce6e-4007-a74b-9f4e7395a02f.png)

**Reply by theyosh** [reply]
So, when you restart it now, there is still a calendar error? What does the command `git status` give you? Are you on the main branch?

**Comment by steveava** [comment]
Yep, been restarted several times.

Git Status:

![image](https://user-images.githubusercontent.com/39738639/232592904-6695525c-6816-4b87-8f91-af5c370e4306.png)

**Reply by theyosh** [reply]
Your code is behind with 19 changes....  Somehow there has been a change in your code... 

What happens when you try `git pull --rebase` will that update more files?

**Comment by steveava** [comment]
Tried git pull --rebase, then git checkout main to get back to the main branch. Tried git pull, still get the same git status response(above screen shot).

Still getting this on a git pull. Haven't seen this in past git pulls.

![image](https://user-images.githubusercontent.com/39738639/232719991-d2daf51e-680d-415c-ba85-9721c1481eda.png)

**Reply by theyosh** [reply]
Hmm, is this still an issue? Did forgot a bit about this
